docs/proposal/2022-11-22-botkube-plugins-integration.md

```diff
@@ -0,0 +1,498 @@
+    description: "Kubectl executor plugin."
+    version: "v1.0.0"
+    links:
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-arm64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-linux-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-linux-arm64
+  - name: "kubernetes"
+    type: "source"
+    description: "Kubernetes source plugin."
+    version: "v1.0.0"
+    links:
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-darwin-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-darwin-arm64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-linux-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-linux-arm64
+```
+
+The index file can be generated automatically just by scanning directory where plugins binaries are stored.
+
+## Implementation details
+
+### E2E testing
+
+Botkube fetches the plugins index and binaries when it's started. However, if they are already present, Botkube doesn't take any action. We could leverage that and simply built all plugins binaries and include them in Docker image with in a proper directories.
+Unfortunately, this approach doesn't test whether the download mechanism works properly. To overcome this issue we can still build all plugins on pull-request but instead of embedding them into Docker images we can start a simple static file server. This server can be later accessed by a running Pod via `host.k3d.internal` DNS which enables access to host system. This feature is natively supported by [k3d](https://k3d.io/v5.0.1/faq/faq/#how-to-access-services-like-a-database-running-on-my-docker-host-machine).
+
+### How plugins are stored
+
+Once downloaded, such folder structure is created:
+
+```plaintext
+cache-dir
+├── botkube                  # repository name
+│  ├── executor_v0.1.0_echo  # executor binary with patter 'executor_<version>_<name>'
+│  └── source_v0.1.0_echo    # source binary with patter 'source_<version>_<name>'
+└── botkube.yaml             # repository index file
+```
+
+Before downloading index file or plugin binaries, Botkube checks whether a given file already exist. If yes, no action is taken.
+
+### Do we want to decouple the interfaces between plugin Go implementation and gRPC?
+
+I don't have a strong opinion here. For now, I reuse the generated struct from the Protocol Buffers.
+
+### Do we want to have a separate go mod for each executor/source?
+
+I don't have a strong opinion here. For now, just for the sake of simplicity I would stay with shared dependencies as long as we don't extract full the `kubectl` executor and `kubernetes` source as external plugins.
+Later we can revisit this decision.
```

Comment:
Agree with both. If the whole team agrees, then let's rephrase the sections to highlight this is a team decision 👍

Reply:
This comment is still not resolved. Please replace the phrases, e.g.:

- Remove "I don't have a strong opinion here"
- Replace "I reuse" to "We'll reuse"
- Replace "I would stay" to "We'll keep the shared dependencies"

Also, replace the questions for the section headers to something like:

- "Decoupling interfaces between Go implementation and gRPC"
- "Separate Go modules for each plugin"